### PR TITLE
Random password

### PR DESCRIPTION
**Pull request by Changron**
1. random public user password
2. redirect to download page after register

### CODE REVIEW

File: contest/contest_info.py [hunk 1]
```diff
@@ -194,8 +199,40 @@ def write_scoreboard_csv_testcases(writer, contest, scoreboard):
...
+        user = contestant.user
+        user.password = make_password(random_password)
+        user.save()
+        logger.info('Public user %s changed password' % (user.username))
+        user_row = [counter+1, user.username,random_password]
+        writer.writerow(user_row)
+
+def get_random_password():
+    #generate random password 
+    #range: A-Z , 0-9
```
Comment (nickelcheng) [inline contest/contest_info.py]: how about A-Z, 0-9, a-z

File: contest/views.py [hunk 2]
```diff
@@ -323,10 +329,29 @@ def reply(request):
...
+        what = request.POST.get('type')
+        if what == 'scoreboard':
+            scoreboard_type = request.POST.get('scoreboard_type')
+            cid = request.POST.get('contest')
+            scoreboard_file = get_scoreboard_csv(cid, scoreboard_type)
+            return scoreboard_file
+        elif what == 'public_user_password':
+            cid = request.POST.get('contest')
+            contest = get_contest_or_404(cid)
+            if user_info.has_contest_ownership(user, contest) or user.has_admin_auth():
```
Comment (nickelcheng) [inline contest/views.py]: This line is too long

File: contest/views.py [hunk 3]
```diff
@@ -323,10 +329,29 @@ def reply(request):
...
 
 def download(request):
-    what = request.POST.get('type')
-    if what == 'scoreboard':
-        scoreboard_type = request.POST.get('scoreboard_type')
-        cid = request.POST.get('contest')
-        scoreboard_file = get_scoreboard_csv(cid, scoreboard_type)
-        return scoreboard_file
-    raise Http404('file not found')
+    user = user_info.validate_user(request.user)
```
Comment (nickelcheng) [inline contest/views.py]: I prefer to use ```@login_required```

Comment (Changron) [inline contest/views.py]: but anyone can download scoreboard
even without logged in

File: contest/contest_info.py [hunk 4]
```diff
@@ -194,8 +199,40 @@ def write_scoreboard_csv_testcases(writer, contest, scoreboard):
...
+    write_public_user_password_csv(writer, contest, public_contestants)
 
+    return response
+
+def write_public_user_password_csv(writer, contest, public_contestants):
+    header = [contest.cname,str(len(public_contestants))+' users']
+    writer.writerow(header)
+    title = ['#','Username','Password']
+    writer.writerow(title)
+    for counter, contestant in enumerate(public_contestants):
```
Comment (nickelcheng) [inline contest/contest_info.py]: I think we don't need a counter here. Only username and password is needed.

**Comment by nickelcheng** [comment]
1. Can you add a back button in download page?
2. I got an UnicodeEncodeError, maybe it's caused by Chinese contest name
screenshot from 2015-04-29 11 30 33
3. Please show warning message when requested public user number is over 200. Don't give users wrong information.
screenshot from 2015-04-29 11 22 15
4. What's the button means? Nothing happened after pushing it.
screenshot from 2015-04-29 11 18 07

File: contest/views.py (hunk 3 above)
Comment (nickelcheng) [inline contest/views.py]: ok

**Comment by nickelcheng** [comment]
I suggest to add a checking alert after the confirm button being pushed.
screenshot from 2015-04-29 11 48 03
To tell user that password will be re-generated and change in public user number(ex: 32 -> 40)

**Comment by Changron** [comment]
'-' is remove in #141 
password regenerated when downloading csv

**Comment by Changron** [comment]
fixed

**Comment by nickelcheng** [comment]
1. When I ask for 0 public user, message shows that "Cannot register public user", but delete all public users
2. I got error when I want to add more public users (I can not reproduce this)
screenshot from 2015-04-29 23 27 43
3. When a user ask for over 200 public users, a warning message should be given instead of a successful message. (I asked for 201 public users in this case)
screenshot from 2015-04-29 23 34 49

4. Can you add current public user number in register dialog? ex: how many public users do you need? you have x public users now

**Comment by Changron** [comment]
fixed

**Comment by nickelcheng** [comment]
Is team 0 a good public user? XD

**Comment by nickelcheng** [comment]
+1 besides the doubt

**Comment by nickelcheng** [comment]
@bruce3557 please check

File: contest/contest_info.py [hunk 5]
```diff
@@ -202,8 +207,40 @@ def write_scoreboard_csv_testcases(writer, contest, scoreboard):
...
+    for contestant in public_contestants:
+        random_password = get_random_password()
+        user = contestant.user
+        user.password = make_password(random_password)
+        user.save()
+        logger.info('Public user %s changed password' % user.username)
+        user_row = [user.username, random_password]
+        writer.writerow(user_row)
+
+def get_random_password():
```
Comment (bruce3557) [inline contest/contest_info.py]: Prefer to put this part in utils because we may reset random password for user?

Comment (bruce3557) [inline contest/contest_info.py]: And this function can be reused

Comment (nickelcheng) [inline contest/contest_info.py]: We won't reset random password for users

Comment (bruce3557) [inline contest/contest_info.py]: I mean forget password option.
If user cannot answer any question, we need to generate a random password?

Comment (nickelcheng) [inline contest/contest_info.py]: Forget password will lead user to its own profile.
The link will send to its email.

**Comment by nickelcheng** [comment]
going to merge.

**Pull request merged**